New reference page for PWA (#25703)
mdn/content: files/en-us/web/progressive_web_apps/reference/index.md

---

> title: Progressive Web Apps reference

Reviewer: Title is interesting here. I guess you chose "Progressive Web Apps reference" rather than just "Reference" so people would get context. But (I think) we will want just "Reference" in the sidebar. We have a front matter key `short-title`, and we could make the PWA sidebar use that if it is set.

Does that sound like a good idea?

Author: Totally agree that just "Reference" would be better for the sidebar. I do think that we need the longer "Progressive Web Apps reference" as the page title because people may stumble on this page from anywhere else on the web, with no prior context. Plus that also means that's what appears in the `<title>` element, so in the browser tab, which I think is important.

The breadcrumbs on MDN should probably also use the longer title, otherwise it would end up as: "References > Progressive web apps (PWAs) > Reference".

The `short-title` idea is nice! I'll add this key on the page.

Reviewer: > The breadcrumbs on MDN should probably also use the longer title, otherwise it would end up as: "References > Progressive web apps (PWAs) > Reference".

I think we will need to use `short-title` in breadcrumbs as well, partly because usually that's what breadcrumbs want, partly because otherwise we get the same repetition in the breadcrumbs, and partly because it's nice for sidebar and breadcrumbs to match. It is a bit unfortunate that we get this "References->Reference" repetition. I think ideally we would not have "References" here at all, and "Progressive Web Apps" would be at the top (or even "Web > Progressive Web Apps"). But I'm not sure how easy that would be to change.

Author: That makes sense! I was focusing on the "Reference" repetition only, but you make great points. Let's probably not worry about the "References" breadcrumbs root in this PR then.

- [`launch_handler`](/en-US/docs/Web/Manifest/launch_handler)
  - : The context in which the app should be loaded when launched.

### Other

- [`scope`](/en-US/docs/Web/Manifest/scope)
  - : The navigation scope of this web application, which restricts what web pages can be viewed while the manifest is applied.
- [`related_applications`](/en-US/docs/Web/Manifest/related_applications)
  - : Alternative applications that provide similar or equivalent functionality to this app. Developers can specify that these alternative applications are preferred by using the `prefer_related_applications` member.
- [`prefer_related_applications`](/en-US/docs/Web/Manifest/prefer_related_applications)
  - : Whether the applications listed in the `related_applications` member should be preferred over this app.
- [`serviceworker`](/en-US/docs/Web/Manifest/serviceworker)
  - : The service worker to use to run a web-based payment app on a merchant website.

## Service Worker APIs

### Communication with the app

The following APIs can be used to communicate between the service worker and the client app:

- [`Client.postMessage()`](/en-US/docs/Web/API/Client/postMessage)

>   - : Allows a service worker to send a message to its client app.

Author: I tried to clarify this by rephrasing the intro line, and by applying your suggestion. Thanks.

- [Broadcast Channel API](/en-US/docs/Web/API/Broadcast_Channel_API)
  - : Allows a service worker and its client app to establish a basic two-way communication channel.

### Offline operation

The following APIs can be used by a service worker to make your app work offline:

- [`Cache`](/en-US/docs/Web/API/Cache)
  - : A persistent storage mechanism for HTTP responses used to store assets that can be reused when the app is offline.
- [`Clients`](/en-US/docs/Web/API/Clients)
  - : An interface used to provide access to the documents that are controlled by the service worker.
- [`FetchEvent`](/en-US/docs/Web/API/FetchEvent)

>   - : An event, dispatched in the service worker with every HTTP request made by a client, that can be used to cache a response or respond with another resource.

Reviewer 2: what is "resource" here? maybe "responsd with a promise", linking promise, but not sure if that is correct.

Author: I'll try to find something. What I want to say here is that the event can be used to either send the request to the server as normal, and then maybe cache the response for next time. Or the event can be used to intercept the request, not even let it go to the server, and respond with another response instead. This could be a previously cached response (or perhaps a newly constructed response object, although that's a less common scenario).

### Background operation

The following APIs can be used by a service worker to perform tasks in the background, even when your app is not running:

- [Background Synchronization API](/en-US/docs/Web/API/Background_Synchronization_API)
  - : A way to defer tasks so that they can be run in a service worker once the user has a stable network connection.
- [Web Periodic Background Synchronization API](/en-US/docs/Web/API/Web_Periodic_Background_Synchronization_API)
  - : A way to register tasks to be run in a service worker at periodic intervals with network connectivity.
- [Background Fetch API](/en-US/docs/Web/API/Background_Fetch_API)

>   - : A method for a service worker to manage downloads that may take a significant amount of time such as video or audio files.

Reviewer 2: ```suggestion
  - : A method for a service worker to manage downloads that may take a significant amount of time, such as video or audio files.
```

## Other web APIs

- [IndexedDB](/en-US/docs/Web/API/IndexedDB_API)
  - : A client-side storage API for significant amounts of structured data, including files.
- [Badging API](/en-US/docs/Web/API/Badging_API)

>   - : A method of setting a badge on the application icon, to act as a low distraction notification.

Reviewer 2: ```suggestion
  - : A method of setting a badge on the application icon, providing a low-distraction notification.
```

- [Notifications API](/en-US/docs/Web/API/Notifications_API)
  - : A way to send notifications that are displayed at the operating system level.
- [Web Share API](/en-US/docs/Web/API/Web_Share_API)
  - : A mechanism for sharing text, links, files, and other content to other apps selected by the user on their device.
- [Window Controls Overlay API](/en-US/docs/Web/API/Window_Controls_Overlay_API)

>   - : An API for PWAs installed on desktop operating systems to hide the default window title bar and display their own content over the full surface area of the app window instead.

Reviewer 2: ```suggestion
  - : An API for PWAs installed on desktop operating systems that enables hiding the default window title bar, enabling displaying the app over the full surface area of the app window.
```

---
title: Progressive Web Apps reference
short-title: Reference
slug: Web/Progressive_web_apps/Reference
page-type: landing-page
---

This reference describes the [web app manifest](/en-US/docs/Web/Manifest) members, [Service Worker APIs](/en-US/docs/Web/API/Service_Worker_API), and other web APIs that [Progressive Web Apps](/en-US/docs/Web/Progressive_web_apps) (PWAs) can use to offer a great user experience.

## Web app manifest members

### Identification

The following manifest members are used to name and describe your PWA:

> - [`name`](/en-US/docs/Web/Manifest/name)

Author: Yeah, agreed on trying to avoid repetition wherever we can, but I'm not sure how. I do think that a PWA reference landing page is useful as it can act as a nice cheat sheet, and gives developer a high level view of everything they can do with PWA.

It'd be kind of cool if we could reuse something like `{{ListSubpages("/en-US/docs/Web/Manifest")}}` but unfortunately this macro only lists the pages, and doesn't provide short descriptions.

I guess, as a way to feel better about this, the list of web app manifest members is actually not repeated on MDN at all right now. This new page would be the only one where it's manually written. It's generated with the macro on the manifest page.

So, sure, it does mean that as soon as there's a new member, not only do we need to document it, but we also need to remember to add it here on this page, but that's it. There are no other pages.

We might even be able to make the case that only PWA-specific members should be listed here anyway. And that's a case-by-case basis. For example, I should probably remove the `serviceworker` member because it's specific to web payment, and not PWA. I believe `id` is unused right now, so perhaps it makes sense to remove it too.
Does that make it easier to accept the maintenance cost?

> - [`name`](/en-US/docs/Web/Manifest/name)

Author: I just pushed a new commit where I changed the list of members a little bit to make it more specific to PWA, as outlined above.

> - [`name`](/en-US/docs/Web/Manifest/name)

Reviewer: I'm just speaking from experience: it will get out of date, and every time we repeat things in MDN is a time we make more work for writers. Each time is tiny, but we do it so much that we spend huge amounts of time fixing everything up in all the places.

But like I say, it's hard to argue against, because repetition is clearly better for readers (at least until it gets out of date), and each individual case seems so small that it's not worth arguing about. 

> It'd be kind of cool if we could reuse something like {{ListSubpages("/en-US/docs/Web/Manifest")}} but unfortunately this macro only lists the pages, and doesn't provide short descriptions.

We could (almost) use [`SubpagesWithSummaries`](https://github.com/mdn/yari/blob/main/kumascript/macros/SubpagesWithSummaries.ejs) which in this case would give us something like this:

<img width="635" alt="Screen Shot 2023-03-30 at 9 34 57 AM" src="https://user-images.githubusercontent.com/432915/228906260-0dcd18c7-1dd9-4150-9415-4789b9efa7c5.png">

We can't quite do that at the moment because apparently `SubpagesWithSummaries` only lets you list subpages of the _current_ page, but it's probably an easy fix to make it accept a parameter here.

> - [`name`](/en-US/docs/Web/Manifest/name)

Reviewer 2: i really like that the members are divided into groups. The full list as a naked list makes sense on https://developer.mozilla.org/en-US/docs/Web/Manifest. This page is about PWA, not manifest, so categories rather than lists are more apropos. 

I think for right now, all current members are relevant to PWA, they should all be included on this page. But in the future that will not be the case

> - [`name`](/en-US/docs/Web/Manifest/name)

Reviewer: Actually, although I don't expect people to agree, I think it is better only to link to the top-level manifest page, because it makes this page much shorter and easier for people to get an overview of the reference docs we have. I think listing all the keys makes this page too long.

> - [`name`](/en-US/docs/Web/Manifest/name)

Author: After some more discussion on this, in another thread, I've decided to follow Will's advice and remove the list of members. Instead, I'll add just one link to the manifest page.

My reasoning is that it's safer to start this way, and we can always add the members back in later, for example using a macro instead.

---
title: Progressive Web Apps reference
short-title: Reference
slug: Web/Progressive_web_apps/Reference
page-type: landing-page
---

This reference describes the technologies, features, and APIs that [Progressive Web Apps](/en-US/docs/Web/Progressive_web_apps) (PWAs) can use to offer a great user experience.

## Web app manifest

- [Web app manifest members](/en-US/docs/Web/Manifest/)

>   - : Developers can use web app manifest members to describe a PWA, customize its appearance, and more deeply integrate it in the operating system.

Reviewer: ```suggestion
  - : Developers can use web app manifest members to describe a PWA, customize its appearance, and more deeply integrate it into the operating system.
```

?
